# Import Libraries

In [2]:
import tensorflow as tf
import pandas as pd

### Import Data

In [3]:
df = pd.read_csv("https://cocl.us/concrete_data")
df.head(2)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89

# Part A 

### Split the data in train and test

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((721, 8), (309, 8), (721,), (309,))

### Create the model

In [7]:
import tensorflow.keras as keras

In [8]:
def build_model():
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(10, activation="relu", input_shape=(X.shape[1],)))
    model.add(keras.layers.Dense(1))
    
    model.compile(optimizer="adam", loss="mean_squared_error")
    return model

In [9]:
model = build_model()

model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 219555.9531
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 102054.0938
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 39985.1992
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 13964.2080
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 5549.1680
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 3622.1892
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 3250.4746
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 3092.8137
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 2940.5669
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 2796.5283
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 2649.2964
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 2503.7070
Epoch 1

### Calculating Mean Squared Error

In [10]:
from sklearn.metrics import mean_squared_error

predicted_y = model.predict(X_test)
mse = mean_squared_error(y_test, predicted_y)
print(f"Test MSE: {mse}")

10/10 [==============================] - 0s 2ms/step
Test MSE: 401.210626918882


In [13]:
def do_fifty_times():
    n=50
    mse_list = []
    for index in range(n):
        # split data
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
        
        # build model
        model = build_model()
        
        # train model
        model.fit(X_train, y_train, epochs=50, verbose=0)
        
        # predict test set
        predicted_y = model.predict(X_test)
        
        # get MSE for the model
        mse = mean_squared_error(y_test, predicted_y)
        print(f"MSE {index}: {mse}")
        
        # append to list
        mse_list.append(mse)
    return mse_list

In [14]:
mse_list = do_fifty_times()

10/10 [==============================] - 0s 2ms/step
MSE 0: 336.9535494107098
10/10 [==============================] - 0s 2ms/step
MSE 1: 104.54002299276166
10/10 [==============================] - 0s 2ms/step
MSE 2: 213.17574254112256
10/10 [==============================] - 0s 2ms/step
MSE 3: 471.4136880042834
10/10 [==============================] - 0s 2ms/step
MSE 4: 119.97814989392786
10/10 [==============================] - 0s 2ms/step
MSE 5: 140.71852562786304
10/10 [==============================] - 0s 2ms/step
MSE 6: 194.67328482940866
10/10 [==============================] - 0s 2ms/step
MSE 7: 151.56492390887544
10/10 [==============================] - 0s 2ms/step
MSE 8: 109.45796481353014
10/10 [==============================] - 0s 2ms/step
MSE 9: 180.53164620221273
10/10 [==============================] - 0s 2ms/step
MSE 10: 667.2925047519243
10/10 [==============================] - 0s 2ms/step
MSE 11: 175.2940958259637
10/10 [==============================] - 0s 2ms/step
M

### Average and Standard Deviation of MSE

In [15]:
import numpy as np
print(f"Average of MSEs: {np.mean(mse_list)}")
print(f"Standard Deviation MSEs: {np.std(mse_list)}")


Average of MSEs: 251.51925597627698
Standard Deviation MSEs: 245.3108361210546


# Part B

# Normalization

In [16]:
df.isna().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [5]:
X = df.drop("Strength", axis=1)
y = df.loc[:, "Strength"]

X.shape, y.shape

((1030, 8), (1030,))

In [17]:
X_norm = (X - X.mean()) / X.std()
X_norm.head(5)

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [15]:
import numpy as np
print(f"Average of MSEs: {np.mean(mse_list)}")
print(f"Standard Deviation MSEs: {np.std(mse_list)}")


Average of MSEs: 251.51925597627698
Standard Deviation MSEs: 245.3108361210546


### Split the data in train and test

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.3)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((721, 8), (309, 8), (721,), (309,))

### Create the model

In [22]:
import tensorflow.keras as keras

In [23]:
def build_model():
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(10, activation="relu", input_shape=(X.shape[1],)))
    model.add(keras.layers.Dense(1))
    
    model.compile(optimizer="adam", loss="mean_squared_error")
    return model

In [24]:
model = build_model()

model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 1611.6862
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1592.4187
Epoch 3/50
23/23 [==============================] - 0s 3ms/step - loss: 1573.2382
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1553.6326
Epoch 5/50
23/23 [==============================] - 0s 3ms/step - loss: 1534.0696
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1513.9550
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1493.0442
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1471.5038
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1449.1288
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1425.8679
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1402.0944
Epoch 12/50
23/23 [==============================] - 0s 3ms/step - loss: 1377.0511
Epoch 13/50
2

### Calculating Mean Squared Error

In [25]:
from sklearn.metrics import mean_squared_error

predicted_y = model.predict(X_test)
mse = mean_squared_error(y_test, predicted_y)
print(f"Test MSE: {mse}")

10/10 [==============================] - 0s 2ms/step
Test MSE: 356.9511484180675


In [26]:
def do_fifty_times():
    n=50
    mse_list = []
    for index in range(n):
        # split data
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
        
        # build model
        model = build_model()
        
        # train model
        model.fit(X_train, y_train, epochs=50, verbose=0)
        
        # predict test set
        predicted_y = model.predict(X_test)
        
        # get MSE for the model
        mse = mean_squared_error(y_test, predicted_y)
        print(f"MSE {index}: {mse}")
        
        # append to list
        mse_list.append(mse)
    return mse_list

In [28]:
mse_list = do_fifty_times()

10/10 [==============================] - 0s 2ms/step
MSE 0: 148.74174397257167
10/10 [==============================] - 0s 2ms/step
MSE 1: 121.31726228924882
10/10 [==============================] - 0s 2ms/step
MSE 2: 126.00204108464314
10/10 [==============================] - 0s 2ms/step
MSE 3: 642.2163175738107
10/10 [==============================] - 0s 2ms/step
MSE 4: 110.84855623046901
10/10 [==============================] - 0s 2ms/step
MSE 5: 101.3624872016314
10/10 [==============================] - 0s 2ms/step
MSE 6: 280.4744785234453
10/10 [==============================] - 0s 2ms/step
MSE 7: 411.84367879027633
10/10 [==============================] - 0s 2ms/step
MSE 8: 106.65094879689423
10/10 [==============================] - 0s 2ms/step
MSE 9: 232.14027720934595
10/10 [==============================] - 0s 2ms/step
MSE 10: 179.68252323698772
10/10 [==============================] - 0s 2ms/step
MSE 11: 163.62413361040038
10/10 [==============================] - 1s 10ms/step

### Average and Standard Deviation of MSE

In [29]:
import numpy as np
print(f"Average of MSEs: {np.mean(mse_list)}")
print(f"Standard Deviation MSEs: {np.std(mse_list)}")


Average of MSEs: 274.0061911191259
Standard Deviation MSEs: 235.82275802587836


# MSE Compared to Part A

For Part A:

    Average of MSEs: 251.51925597627698
    Standard Deviation MSEs: 245.3108361210546
    
For Part B:

    Average of MSEs: 274.0061911191259
    Standard Deviation MSEs: 235.82275802587836
    
As we can see that Average MSE has increased while Standard Deviation remains the same

# Part C

In [30]:
import tensorflow.keras as keras

In [31]:
def build_model():
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(10, activation="relu", input_shape=(X.shape[1],)))
    model.add(keras.layers.Dense(1))
    
    model.compile(optimizer="adam", loss="mean_squared_error")
    return model

In [32]:
model = build_model()

model.fit(X_train, y_train, epochs=100, verbose=1)

Epoch 1/100
23/23 [==============================] - 1s 2ms/step - loss: 1594.5854
Epoch 2/100
23/23 [==============================] - 0s 2ms/step - loss: 1576.6569
Epoch 3/100
23/23 [==============================] - 0s 3ms/step - loss: 1559.0067
Epoch 4/100
23/23 [==============================] - 0s 3ms/step - loss: 1541.0098
Epoch 5/100
23/23 [==============================] - 0s 3ms/step - loss: 1522.7096
Epoch 6/100
23/23 [==============================] - 0s 3ms/step - loss: 1504.1951
Epoch 7/100
23/23 [==============================] - 0s 2ms/step - loss: 1484.9303
Epoch 8/100
23/23 [==============================] - 0s 3ms/step - loss: 1465.1403
Epoch 9/100
23/23 [==============================] - 0s 3ms/step - loss: 1444.8527
Epoch 10/100
23/23 [==============================] - 0s 2ms/step - loss: 1423.8485
Epoch 11/100
23/23 [==============================] - 0s 3ms/step - loss: 1402.0864
Epoch 12/100
23/23 [==============================] - 0s 3ms/step - loss: 1380.1022
E

23/23 [==============================] - 0s 2ms/step - loss: 164.1960
Epoch 100/100
23/23 [==============================] - 0s 2ms/step - loss: 162.9955


### Calculating Mean Squared Error

In [33]:
from sklearn.metrics import mean_squared_error

predicted_y = model.predict(X_test)
mse = mean_squared_error(y_test, predicted_y)
print(f"Test MSE: {mse}")

10/10 [==============================] - 0s 2ms/step
Test MSE: 164.66211049293042


In [34]:
def do_fifty_times():
    n=50
    mse_list = []
    for index in range(n):
        # split data
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
        
        # build model
        model = build_model()
        
        # train model
        model.fit(X_train, y_train, epochs=50, verbose=0)
        
        # predict test set
        predicted_y = model.predict(X_test)
        
        # get MSE for the model
        mse = mean_squared_error(y_test, predicted_y)
        print(f"MSE {index}: {mse}")
        
        # append to list
        mse_list.append(mse)
    return mse_list

In [35]:
mse_list = do_fifty_times()

10/10 [==============================] - 0s 2ms/step
MSE 0: 148.25104507749646
10/10 [==============================] - 0s 2ms/step
MSE 1: 440.8798405042123
10/10 [==============================] - 0s 2ms/step
MSE 2: 150.40248650672098
10/10 [==============================] - 0s 2ms/step
MSE 3: 173.04908351539135
10/10 [==============================] - 0s 2ms/step
MSE 4: 122.90174160117273
10/10 [==============================] - 0s 2ms/step
MSE 5: 130.6660630796108
10/10 [==============================] - 0s 2ms/step
MSE 6: 1377.6815908284989
10/10 [==============================] - 0s 2ms/step
MSE 7: 465.99601849541574
10/10 [==============================] - 0s 2ms/step
MSE 8: 139.1456254079793
10/10 [==============================] - 0s 2ms/step
MSE 9: 2477.8960647333165
10/10 [==============================] - 0s 2ms/step
MSE 10: 151.47762327719803
10/10 [==============================] - 0s 2ms/step
MSE 11: 279.9292423387407
10/10 [==============================] - 0s 2ms/step
M

### Average and Standard Deviation of MSE

In [36]:
import numpy as np
print(f"Average of MSEs: {np.mean(mse_list)}")
print(f"Standard Deviation MSEs: {np.std(mse_list)}")


Average of MSEs: 442.814296697479
Standard Deviation MSEs: 526.5835180443946


# MSE Compared to Part B

For Part B:

        Average of MSEs: 274.0061911191259
        Standard Deviation MSEs: 235.82275802587836

For Part C:

        Average of MSEs: 442.814296697479
        Standard Deviation MSEs: 526.5835180443946

Both Mean MSE and SD have increased

# Part D

In [37]:
import tensorflow.keras as keras

In [38]:
def build_model():
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(10, activation="relu", input_shape=(X_norm.shape[1],)))
    model.add(keras.layers.Dense(10, activation="relu"))
    model.add(keras.layers.Dense(10, activation="relu"))
    model.add(keras.layers.Dense(1))
    
    model.compile(optimizer="adam", loss="mean_squared_error")
    return model

In [39]:
model = build_model()

model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 3s 2ms/step - loss: 1602.8252
Epoch 2/50
23/23 [==============================] - 0s 3ms/step - loss: 1576.4741
Epoch 3/50
23/23 [==============================] - 0s 3ms/step - loss: 1546.6514
Epoch 4/50
23/23 [==============================] - 0s 3ms/step - loss: 1507.8400
Epoch 5/50
23/23 [==============================] - 0s 3ms/step - loss: 1452.8308
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1367.0994
Epoch 7/50
23/23 [==============================] - 0s 3ms/step - loss: 1233.0011
Epoch 8/50
23/23 [==============================] - 0s 3ms/step - loss: 1057.6328
Epoch 9/50
23/23 [==============================] - 0s 3ms/step - loss: 861.0961
Epoch 10/50
23/23 [==============================] - 0s 3ms/step - loss: 666.7316
Epoch 11/50
23/23 [==============================] - 0s 3ms/step - loss: 515.5941
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 412.4381
Epoch 13/50
23/23

### Calculating Mean Squared Error

In [40]:
from sklearn.metrics import mean_squared_error

predicted_y = model.predict(X_test)
mse = mean_squared_error(y_test, predicted_y)
print(f"Test MSE: {mse}")

10/10 [==============================] - 0s 2ms/step
Test MSE: 137.2090936257836


In [41]:
def do_fifty_times():
    n=50
    mse_list = []
    for index in range(n):
        # split data
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
        
        # build model
        model = build_model()
        
        # train model
        model.fit(X_train, y_train, epochs=50, verbose=0)
        
        # predict test set
        predicted_y = model.predict(X_test)
        
        # get MSE for the model
        mse = mean_squared_error(y_test, predicted_y)
        print(f"MSE {index}: {mse}")
        
        # append to list
        mse_list.append(mse)
    return mse_list

In [42]:
mse_list = do_fifty_times()

10/10 [==============================] - 0s 2ms/step
MSE 0: 91.60750640460417
10/10 [==============================] - 0s 2ms/step
MSE 1: 117.46177724344687
10/10 [==============================] - 0s 2ms/step
MSE 2: 116.63984072170814
10/10 [==============================] - 0s 2ms/step
MSE 3: 98.63364276310654
10/10 [==============================] - 0s 2ms/step
MSE 4: 116.81053708741193
10/10 [==============================] - 0s 2ms/step
MSE 5: 233.34644255560846
10/10 [==============================] - 0s 2ms/step
MSE 6: 72.43754200544481
10/10 [==============================] - 0s 2ms/step
MSE 7: 148.58248224801855
10/10 [==============================] - 0s 2ms/step
MSE 8: 71.51021913380364
10/10 [==============================] - 0s 2ms/step
MSE 9: 124.80762659556004
10/10 [==============================] - 0s 2ms/step
MSE 10: 98.11132386517083
10/10 [==============================] - 0s 2ms/step
MSE 11: 127.58064742170461
10/10 [==============================] - 0s 2ms/step
MS

### Average and Standard Deviation of MSE

In [43]:
import numpy as np
print(f"Average of MSEs: {np.mean(mse_list)}")
print(f"Standard Deviation MSEs: {np.std(mse_list)}")


Average of MSEs: 122.78204620575117
Standard Deviation MSEs: 46.05085031127403


# MSE Compared to Part B

For part B:
        
        Average of MSEs: 274.0061911191259
        Standard Deviation MSEs: 235.82275802587836

For Part D:
    
        Average of MSEs: 122.78204620575117
        Standard Deviation MSEs: 46.05085031127403
            
Both Average MSE and Standard Deviation has decreased, hence accuracy has increased